### ISSUE TITLE
KinematicBody2D performance degradation

### ISSUE THREAD

**Issue by mephiticstudios** [godotengine/godot#19579]
**Godot version:**
3.0.3 rc1+ possibly sooner. 

**OS/device including version:**
Ubuntu Linux 

**Issue description:**
 move_and_collide** / move_and_slide take longer and longer to execute until fps drop to 1. 
- When the rotation of the body is at a multiple of 90 degrees the performance goes back to normal,
- The closer to a 45 degree angle the body rotates. the slower it gets. 
- The performance degradation also seems to be a function of how far the body is from its origin. 

**Steps to reproduce:**
After significant debugging I've nailed it down to only occurring when:
- **the KinematicBody2D has 2 collision shapes and one of them is disabled.** 
- I used a one Capsule and only CollisionPolygon2D but it may not matter.  If both shapes on or both shapes are off the performance is not impacted.

**Comment by mephiticstudios** [comment]
Yes this occurs in master, I kept recompiling over the week hoping it was some bug that was already known.

Degradation wasn't the best choice of words. Basically once you get in the pathological case, the time for each move_and_collide increases on each iteration until the project eventually chugs at less than 1 fps. Even in the worst case rotating back to a 90 degree angle "unsticks it" and rotation back immediate stops it again. 

If I had to guess it would be that some calculation is still taking the disabled shape into account and maybe thinks its still located at the origin? 

 I only recently started disabling a collision shape so I'm uncertain how far back this might go.

**Comment by akien-mga** [comment]
If you can provide a zipped project that reproduces the issue, that would greatly help with debugging. Making sure that you free nodes (`queue_free()`) when you want to get rid of them, if you just remove them from the tree `remove_child()` the memory is not freed.

**Comment by mephiticstudios** [comment]
Sure thing, I've attached a project.Use arrow keys to turn and move, if you fly in any direction for 5-10 seconds and are rotated in certain ways your performance should fall off a cliff. Flying back to origin or rotating to a multiple of 90 degree's should improve it. 

I'm not destroying shapes, I'm just disabling one. 

[reproduce_issue.zip](https://github.com/godotengine/godot/files/2108513/reproduce_issue.zip)

**Comment by guilhermefelipecgs** [comment]
Confirmed. The problem seens to be in `move_and_collide` and `move_and_slide`.
